library-api/src/main/java/au/csiro/pathling/library/PathlingContext.java

```diff
@@ -11,31 +11,70 @@
  *
  */
 
-package au.csiro.pathling.api;
+package au.csiro.pathling.library;
 
+import static au.csiro.pathling.fhirpath.encoding.SimpleCodingsDecoders.COL_ARG_CODINGS;
+import static au.csiro.pathling.fhirpath.encoding.SimpleCodingsDecoders.COL_INPUT_CODINGS;
+import static au.csiro.pathling.utilities.Preconditions.wrapInUserInputError;
 import static java.util.Objects.nonNull;
+import static org.apache.spark.sql.functions.array;
+import static org.apache.spark.sql.functions.lit;
+import static org.apache.spark.sql.functions.struct;
+import static org.apache.spark.sql.functions.when;
 
 import au.csiro.pathling.encoders.FhirEncoders;
+import au.csiro.pathling.fhir.DefaultTerminologyServiceFactory;
+import au.csiro.pathling.fhir.TerminologyServiceFactory;
+import au.csiro.pathling.fhirpath.encoding.CodingEncoding;
+import au.csiro.pathling.fhirpath.encoding.SimpleCoding;
+import au.csiro.pathling.fhirpath.encoding.SimpleCodingsDecoders;
+import au.csiro.pathling.fhirpath.function.memberof.MemberOfMapper;
+import au.csiro.pathling.fhirpath.function.subsumes.SubsumesMapper;
+import au.csiro.pathling.fhirpath.function.translate.TranslateMapper;
+import au.csiro.pathling.sql.MapperWithPreview;
+import au.csiro.pathling.sql.PathlingStrategy;
+import au.csiro.pathling.sql.SqlExtensions;
 import au.csiro.pathling.support.FhirConversionSupport;
+import au.csiro.pathling.terminology.ConceptTranslator;
+import au.csiro.pathling.terminology.Relation;
+import au.csiro.pathling.utilities.Strings;
+import ca.uhn.fhir.context.FhirContext;
 import ca.uhn.fhir.context.FhirVersionEnum;
 import ca.uhn.fhir.context.RuntimeResourceDefinition;
 import java.util.List;
+import java.util.Set;
+import java.util.UUID;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.spark.sql.Column;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Encoders;
 import org.apache.spark.sql.Row;
 import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
 import org.hl7.fhir.instance.model.api.IBaseBundle;
 import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Enumerations.ConceptMapEquivalence;
+import org.slf4j.MDC;
 
 /**
- * Main entry point to Pathling Java API
+ * A class designed to provide access to selected Pathling functionality from non-JVM language
+ * libraries.
+ *
+ * @author Piotr Szul
+ * @author John Grimes
  */
 public class PathlingContext {
 
```

Comment:
I am wondering if we actually need the `PathlingContext` in this form both in Java and in Python.
Firstly we do not actually need to store SparkSession because all our functions take a dataframe as an argument an we could get the session from it.
Secondly maybe it would be better to split this into to "contexts": one specialised for encoding (only needing the relevant argument) and one specialized terminology.

maybe something like this:

``` 
import pathling.Encoder
import pathling.Terminlogy

encoder = Encoder("R4", max_nesting_level = 2)
encoder.encode_json( ....)


term = Temininology("http://server.url).
term.member_of (....)

```

Reply:
I agree that we shouldn't need to store the `SparkSession` in the `PathlingContext`. I have removed it from the Java class here: a676d227b5232220bb00605eb3dc3569014a8c28

Not too sure about making the same change in the Python `PathlingContext`, as that might impact your implementation of `PathlingContext._wrap_df`.

I think I might need some more convincing about the "different contexts for different use cases" thing - I think that might make things more complicated. I'm pretty happy with the fact that we can instantiate this thing and use it for either use case, all of the arguments are optional.

I can see a case for a more sophisticated solution for passing configuration to the PathlingContext as we build more functionality into it. The named arguments feature of Python means that this is not so bad even as it is, I think.